### PR TITLE
chore: added function calls proposal

### PR DESCRIPTION
**Pull request by otaviomacedo**
Proposals for function call syntax in deCDK. We'd love to get feedback from the community on what their preferred syntax is, things we have missed, new ideas etc.

### CODE REVIEW

File: docs/function-calls-syntax.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+Inline:
+
+```yaml
+Resources:
+  MyLambda:
+    Type: aws-cdk-lib.aws_lambda.Function
+    Properties:
+      # ...
+  MyDataSource:
+    Type: aws-cdk-lib.aws_appsync..LambdaDataSource
```
Comment (mrgrain) [inline docs/function-calls-syntax.md:90]:
```suggestion
    Type: aws-cdk-lib.aws_appsync.LambdaDataSource
```

**Review by mrgrain — CHANGES_REQUESTED** [review]

**Review by mrgrain — DISMISSED** [review]
LGTM. Added `do-not-merge` label in case you'd like @jlhood to have another look.

**Comment by mrgrain** [comment]
```yaml
Resources:
  MyAccessPoint:
    Type: AWSCDK::Call
    Target:
      Ref: MyFileSystem # this can also be 
        Function: 'AddAccessPoint'
        Arguments:
            - 'AccessPoint'
```

```yaml
Resources:
  MyBucket:
    Type: AWSCDK::Call
    Target: aws-cdk-lib.aws_s3.Bucket
    Target:
      Fqn: aws-cdk-lib.aws_s3.Bucket
    Function: 'fromBucketName'
    Arguments:
      - 'AccessPoint'

  MyBucketDeployment:
    Type: aws-cdk-lib.aws_s3_deployment.BucketDeployment
    Properties:
      bucket: 
        aws-cdk-lib.aws_s3.Bucket.fromBucketName: my-bucket
```

**Comment by rix0rrr** [comment]
Here are some more possibilities we talked about:

```yaml


  SomeComplexValue:
    Value: { 'Cdk::Call': { aws-cdk-lib.Duration.seconds: 5 } }
    ObviousContext: { aws-cdk-lib.Duration.seconds: 5 }  # <-- common scenario, therefore lightweight syntax


Resources:
  BlaBucket:
    Type: aws-cdk-lib.aws_s3.Bucket
    Properties:
      SomeThingTypedAsAny: { 'Cdk::Call': { aws-cdk-lib.aws_s3.Bucket.grantRead: [{ Ref: MyBucket }, 'files/*'] } } # <-- disallow!
      SomeThingTypedAsAny: { aws-cdk-lib.aws_s3.Bucket.grantRead: [{ Ref: MyBucket }, 'files/*'] }  

  # Member function 1
  BucketPermissions:
    Type: aws-cdk-lib.aws_iam.Grant
    On: { Ref: BlaBucket }
    Call: { grantRead: 'files/*' }
      
  # Member function 2
  BucketPermissions:
    Type: aws-cdk-lib.aws_iam.Grant
    Call: { aws-cdk-lib.aws_s3.Bucket.grantRead: [{ Ref: MyBucket }, 'files/*'] } # <-- Python/C++ like unbound member functions
    Call: { aws-cdk-lib.aws_s3.Bucket.grantRead: ['MyBucket', 'files/*'] }

  # Member function 3
  BucketPermissions:
    Type: aws-cdk-lib.aws_iam.Grant
    Call: { MyBucket.grantRead: 'files/*' }

  # Static function 1
  BucketPermissions:
    Type: aws-cdk-lib.aws_iam.Grant
    On: aws-cdk-lib.aws_s3.Bucket
    Call: { fromBucketName: 'files/*' }

  # Static function 2
  BucketPermissions:
    # Type: aws-cdk-lib.aws_s3.Bucket (optional?)
    Call: { aws-cdk-lib.aws_s3.Bucket.fromBucketName: { Ref: MyL1Bucket } }

```

**Comment by rix0rrr** [comment]
About scope and ID:

```
  BucketPermissions:
    # Type: aws-cdk-lib.aws_s3.Bucket (optional?)
    Call: { aws-cdk-lib.aws_s3.Bucket.fromBucketName: { Ref: MyL1Bucket } }
    Call: { aws-cdk-lib.aws_s3.Bucket.fromBucketName: { 'Cdk::Args': [{ Ref: 'CDK::Scope' }, 'BucketPermissions', { Ref: MyL1Bucket }] } }

```

**Comment by mrgrain** [comment]
```yaml
 # Member function 4
  BucketPermissions:
    Type: aws-cdk-lib.aws_iam.Grant
    Call:
      - MyBucket 
      - grantRead: 'files/*' 
```

**Pull request closed without merging**